File: src/main/java/prototype/deep/Employee.java
package prototype.deep;

// Deep Copy of an Object
public class Employee implements Cloneable {

    private int empoyeeId;
    private String employeeName;
    private Department department;

    public Employee(int id, String name, Department dept) {
        this.empoyeeId = id;
        this.employeeName = name;
        this.department = dept;
    }

    @Override
    protected Object clone() throws CloneNotSupportedException {
        Employee cloned = (Employee) super.clone();
        cloned.setDepartment((Department) cloned.getDepartment().clone());
        return cloned;
    }

    public int getEmpoyeeId() {
        return empoyeeId;
    }

    public void setEmpoyeeId(int empoyeeId) {
        this.empoyeeId = empoyeeId;
    }

    public String getEmployeeName() {
        return employeeName;
    }

    public void setEmployeeName(String employeeName) {
        this.employeeName = employeeName;
    }

    public Department getDepartment() {
        return department;
    }

    public void setDepartment(Department department) {
        this.department = department;
    }

    @Override
    public String toString() {
        return "Employee{" +
                "empoyeeId=" + empoyeeId +
                ", employeeName='" + employeeName + '\'' +
                ", department=" + department +
                '}';
    }
}

File: src/main/java/prototype/kotlin/Employee.kt
package prototype.kotlin

fun main(args: Array<String>) {

    var department = Department("IT")
    var employee = Employee(1, "Bob", department)
    var otheEmployee = employee.copy(id = 2, employeeName = "Mark")
    var anotherEmployee = employee.copy()
    println(employee)
    println(otheEmployee)
    println(anotherEmployee)

}

data class Employee(val id: Int, val employeeName: String, val department: Department)
data class Department(val name: String)